contract/r/gnoswap/v1/staker/reward_calculation_tick.gno

```diff
@@ -142,46 +144,159 @@ func (self *Tick) updateCurrentOutsideAccumulation(timestamp int64, acc *u256.Ui
+		func(cur realm) {
+			pool, ok := pools.Get(poolPath)
+			if !ok {
+				return
+			}
 
-		var nextTick int32
-		if zeroForOne {
-			nextTick = tickId - 1
-		} else {
-			nextTick = tickId
-		}
+			// Initialize batch processor
+			currentSwapBatch = &SwapBatchProcessor{
+				poolPath:  poolPath,
+				pool:      pool,
+				crosses:   make([]SwapTickCross, 0), // Pre-allocate for typical swap
+				timestamp: timestamp,
+				isActive:  true,
+			}
+		}(cross)
+	}
+}
 
-		liquidityInRangeDelta := tick.stakedLiquidityDelta
-		if liquidityInRangeDelta.Sign() == 0 {
-			return
-		}
+// swapEndHook is called when a swap ends
+func swapEndHook(pools *Pools) func(poolPath string) {
+	return func(poolPath string) {
+		func(cur realm) {
+			if currentSwapBatch == nil || !currentSwapBatch.isActive || currentSwapBatch.poolPath != poolPath {
+				return
+			}
 
-		if zeroForOne {
-			liquidityInRangeDelta = i256.Zero().Neg(liquidityInRangeDelta)
-		}
-		newAcc := pool.modifyDeposit(liquidityInRangeDelta, timestamp, nextTick)
+			// Mark as inactive to prevent further accumulation
+			currentSwapBatch.isActive = false
+
+			// Process all accumulated tick crosses
+			processBatchedTickCrosses()
+
+			// Clean up
+			currentSwapBatch = nil
+		}(cross)
+	}
+}
+
+
+// tickCrossHook is called when a tick is crossed
+// Optimized to batch process tick crosses during swaps
+func tickCrossHook(pools *Pools) func(poolPath string, tickId int32, zeroForOne bool, timestamp int64) {
+	return func(poolPath string, tickId int32, zeroForOne bool, timestamp int64) {
+		func(cur realm) {
+			pool, ok := pools.Get(poolPath)
+			if !ok {
+				return
+			}
+
+			tick := pool.ticks.Get(tickId)
+			if tick.stakedLiquidityDelta.Sign() == 0 {
+				return
+			}
+
+			// If batch processor is active for this pool, accumulate
+			if currentSwapBatch != nil && currentSwapBatch.isActive && currentSwapBatch.poolPath == poolPath {
+				liquidityDelta := tick.stakedLiquidityDelta
+				if zeroForOne {
+					liquidityDelta = i256.Zero().Neg(liquidityDelta)
+				}
+
+				currentSwapBatch.crosses = append(currentSwapBatch.crosses, SwapTickCross{
+					tickId:     tickId,
+					zeroForOne: zeroForOne,
+					delta:      liquidityDelta,
+				})
+				return
+			}
+
+			// Fallback: process immediately (for non-swap tick crosses)
+			processTickCrossImmediate(pool, tick, tickId, zeroForOne, timestamp)
+		}(cross)
+	}
+}
+
+// processTickCrossImmediate processes a single tick cross immediately
+func processTickCrossImmediate(pool *Pool, tick *Tick, tickId int32, zeroForOne bool, timestamp int64) {
+	nextTick := tickId
+	if zeroForOne {
+		nextTick--
+	}
+
+	liquidityDelta := tick.stakedLiquidityDelta
+	if zeroForOne {
+		liquidityDelta = i256.Zero().Neg(liquidityDelta)
+	}
+
+	newAcc := pool.modifyDeposit(liquidityDelta, timestamp, nextTick)
+	tick.updateCurrentOutsideAccumulation(timestamp, newAcc)
+}
+
+// processBatchedTickCrosses processes all accumulated tick crosses at once
```

Comment:
I think that if a failure occurs in the middle of batch processing, only part of the data might be updated. worth consider to add rollback system.

Reply:
additionally, this is a minor point, but this function appears to be the main logic, it would be good to add detailed context to the comments rather than just simple explanations